### ISSUE TITLE
Create a conversational bot/GH Action to handle releases

### ISSUE THREAD

**Issue by gsmet** [quarkusio/quarkus#36285]
I thought about automating releases a lot lately and one of the issue I had is that it's not very easy to handle proper input and proper feedback with a manual workflow run.

I was wondering if we could create something (either a bot or a GH Action - for now, I was more thinking about a GH Action that would react to issue events so that we don't rely on the bot being available) that would actually handle the release and guide you through it with proper error feedback.

My idea was to have the following workflow:
- In a specific project, in a specific org (see notes below), we would create a new issue following a (rather strict) template
- The template would contain everything you need to know about the input variables
- We would check if the issue author is in a specific team of the org, if not, we would error out and close the issue right away
- Depending on the input variables, we would check things, infer the version to release, make sure everything is set properly and add a comment asking the user to confirm the release (the comment would contain all the inferred details)
- The OP (or another person from the aforementioned group) would have to confirm
- Then we would start the first step of the workflow (the release-prepare one)
- Once done, we would add a comment (if failure, we would add a comment too, suggesting a retry)
- The release-perform workflow would then be started automatically
- Once done, we would add a comment (if failure, we would add a comment too, suggesting a retry)
- and so on

Some steps will be a bit harder to handle. For instance, after the release-perform succeeded, we would have to wait ~1 hour and make sure the artifacts are available on Maven Central. Also, the update of the Platform will probably still require some manual intervention (especially if we want to wait for CI to be green there) but still the manual steps could be handled separately and then get back to the issue and go through the additional steps.

I'm pretty sure it's going to be a bumpy road but I think it's worth a try and it could be refined along the way.

Also, we could start with some steps and add more when they are ready.

Notes:
- it's better to do it in a specific org so that we are not blocked by the QuarkusIO org limits, we will have to use a PAT to push changes to the repositories
- one important adjustment is that we will need to get the repository content using `actions/checkout` to make sure the PAT is properly handled (and same for when the doc is synced to the website)
- we need to keep the manual workflow working so the idea is just to wire the existing scripts. And if they need adjustments, they need to keep working in manual mode
- there is a good chance we will need to keep the workspace around while releasing, we can upload an artifact containing the current namespace, the artifact would have the issue number in its name to make sure we can parallelize releases
- retrying will probably require some additional work but it's definitely worth a try given how many problems can happen during the release
- we need some sort of state machine to make sure we know the current state
- the state needs to be stored in the issue as a comment

/cc @maxandersen @gastaldi @aloubyansky for initial feedback

**Comment by gsmet** [comment]
Also /cc @yrodiere as he might have some interesting ideas

**Comment by yrodiere** [comment]
This is interesting, but did you consider simply using a different CI to perform the releases? It seems like a lot of the things you intend to do are just to compensate missing features in GitHub Actions.

E.g. Jenkins allows interactive steps and saving the workspace to restore it later, which could eliminate a lot of problems...

Granted, Jenkins comes with its own set of problems :)

**Comment by gsmet** [comment]
> Granted, Jenkins comes with its own set of problems :)

That exactly.
I think we could come up with something working well and not overly complex while being in a CI environment that is consistent with the rest of our infra.

I think it's worth the experiment and if too complex, we can just abort.

**Comment by gastaldi** [comment]
> This is interesting, but did you consider simply using a different CI to perform the releases? It seems like a lot of the things you intend to do are just to compensate missing features in GitHub Actions.
> 
> E.g. Jenkins allows interactive steps and saving the workspace to restore it later, which could eliminate a lot of problems...

This is a good idea, and we can integrate it with our Red Hat SSO for additional security. The interactive steps (Pipeline) in Jenkins sounds like what you're looking for. Not to mention that is easier to test compared to GH Actions.

> I think we could come up with something working well and not overly complex while being in a CI environment that is consistent with the rest of our infra.

What is exactly needed in our infra to release? I thought only having JDK 11 + Maven + Gradle would be enough, but of course, I'm just seeing the big picture here.

**Comment by gsmet** [comment]
> What is exactly needed in our infra to release?

I meant I would prefer not having a Jenkins instance to update/monitor. And Jenkins pipelines are a pain to write and maintain in my experience.

The idea there was to have Java code triggered by GH Actions to handle most things and delegate to the existing release scripts.

Also with notifications handled automatically by GH as we would add comments.

My experience with Jenkins is that it's a time sink and I'm especially worried about long term maintenance of these pipelines (my experience is that it's not easy to get them right and even harder to get back to them every 6 months when you need an adjustment).

I might be mistaken but I was expecting this to be pretty straightforward with GH Actions.

**Comment by gsmet** [comment]
I started working on this:

- https://github.com/quarkusio/conversational-release-action
- https://github.com/quarkus-release/release/

I'm still experimenting with it to determine if it's viable.

**Comment by maxandersen** [comment]
I Think its interesting to explore too. 

My only real concern is how This Will relate to finding ways to do builds/release for embargoed CVEs. Would this mechanism then be worth it or make it harder?

**Comment by gsmet** [comment]
This is now used as our standard release process for some time.

(And it's awesome!)
